File: leetcode/c/226.翻转二叉树.c
/*
 * @lc app=leetcode.cn id=226 lang=c
 *
 * [226] 翻转二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

// Iterative
struct TreeNode* invertTree(struct TreeNode* root){
    if (root == NULL) {
      return NULL;
    }
    struct TreeNode* stack[1000];
    int top = 0;
    stack[top++] = root;
    while (top > 0) {
      struct TreeNode* node = stack[--top];
      struct TreeNode* tmp = node->left;
      node->left = node->right;
      node->right = tmp;
      if (node->left != NULL) {
        stack[top++] = node->left;
      }
      if (node->right != NULL) {
        stack[top++] = node->right;
      }
    }
    return root;
}
// Recursive
// struct TreeNode* invertTree(struct TreeNode* root){
//   if (root == NULL) {
//     return NULL;
//   }
//   struct TreeNode* tmp = root->left;
//   root->left = root->right;
//   root->right = tmp;
//   invertTree(root->left);
//   invertTree(root->right);
//   return root;
// }
// https://medium.com/@mefengl/comparing-recursive-and-iterative-approaches-to-invert-a-binary-tree-9809dc800e66
// @lc code=end

File: leetcode/ts/226.翻转二叉树.ts
/*
 * @lc app=leetcode.cn id=226 lang=typescript
 *
 * [226] 翻转二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

// Recursive
function invertTree(root: TreeNode | null): TreeNode | null {
  if (!root) return null;
  [root.left, root.right] = [invertTree(root.right), invertTree(root.left)]
  return root
};

// Interation
// function invertTree(root: TreeNode | null): TreeNode | null {
//   if (!root) return null;
//   const stack = [root]
//   while (stack.length) {
//     const node = stack.pop()
//     if (node) {
//       [node.left, node.right] = [node.right, node.left]
//       node.left && stack.push(node.left)
//       node.right && stack.push(node.right)
//     }
//   }
//   return root
// };

// @lc code=end
